packages/react-router/src/router.ts

```diff
@@ -1797,91 +1779,146 @@ export class Router<
 
                 err.routerCode = routerCode
                 firstBadMatchIndex = firstBadMatchIndex ?? index
-                handleRedirectAndNotFound(match, err)
+                handleRedirectAndNotFound(getMatch(), err)
 
                 try {
                   route.options.onError?.(err)
                 } catch (errorHandlerErr) {
                   err = errorHandlerErr
-                  handleRedirectAndNotFound(match, err)
+                  handleRedirectAndNotFound(getMatch(), err)
                 }
 
-                matches[index] = match = updateMatch(match.id, () => ({
-                  ...match,
+                updateMatch(matchId, (prev) => ({
+                  ...prev,
                   error: err,
                   status: 'error',
                   updatedAt: Date.now(),
                   abortController: new AbortController(),
                 }))
               }
 
-              if (match.paramsError) {
-                handleSerialError(match.paramsError, 'PARSE_PARAMS')
-              }
+              const runBeforeLoad = async () => {
+                if (getMatch().beforeLoadPromise) {
+                  return await getMatch().beforeLoadPromise!
```

Comment:
remove `!`

Reply:
TypeScript still thinks it's necessary though, since it *could* have changed somehow during the second `getMatch` call. I hate this about TS, and it's one of the annoyances with using TS with Solid.

Reply:
I just noticed this when testing the PR locally... well then we'll keep it